modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala

```diff
@@ -1377,61 +1393,89 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
+  ): TypedNode[NodeAndType] = {
+    def notSupported(nodeAndPrimitive: (Node, Primitive)) =
+      throw new NotImplementedError(
+        s"Unsupported case: $nodeAndPrimitive"
       )
-    case (_, Primitive.Unit) =>
-      TypedNode.PrimitiveTN(
-        Primitive.Unit,
-        ()
-      )
-    case other =>
-      throw new NotImplementedError(s"Unsupported case: $other")
+    (node, p) match {
+      // String
+      case (N.StringNode(str), Primitive.String) =>
+        TypedNode.PrimitiveTN(Primitive.String, str)
+      // Numeric
+      case (N.NumberNode(num), Primitive.Int) =>
+        TypedNode.PrimitiveTN(Primitive.Int, num.intValue())
+      case (N.NumberNode(num), Primitive.Long) =>
+        TypedNode.PrimitiveTN(Primitive.Long, num.longValue())
+      case (N.NumberNode(num), Primitive.Double) =>
+        TypedNode.PrimitiveTN(Primitive.Double, num.doubleValue())
+      case (N.NumberNode(num), Primitive.Float) =>
+        TypedNode.PrimitiveTN(Primitive.Float, num.floatValue())
+      case (N.NumberNode(num), Primitive.Short) =>
+        TypedNode.PrimitiveTN(Primitive.Short, num.shortValue())
+      case (N.NumberNode(num), Primitive.BigDecimal) =>
+        TypedNode.PrimitiveTN(
+          Primitive.BigDecimal,
+          BigDecimal(num.doubleValue())
+        )
+      case (N.NumberNode(num), Primitive.BigInteger) =>
+        TypedNode.PrimitiveTN(Primitive.BigInteger, BigInt(num.intValue()))
+      // Boolean
+      case (N.BooleanNode(bool), Primitive.Bool) =>
+        TypedNode.PrimitiveTN(Primitive.Bool, bool)
+      case (node, Primitive.Document) =>
+        TypedNode.PrimitiveTN(Primitive.Document, node)
+      case (node, Primitive.String) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.String, "")
+      case (node, Primitive.Int) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Int, 0)
+      case (node, Primitive.Long) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Long, 0L)
+      case (node, Primitive.Double) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Double, 0.0)
+      case (node, Primitive.Float) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Float, 0.0f)
+      case (node, Primitive.Short) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Short, 0: Short)
+      case (node, Primitive.Byte) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Byte, 0.toByte)
+      case (node, Primitive.Blob) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Blob, Array.empty[Byte])
+      case (node, Primitive.Bool) if node == Node.nullNode =>
+        TypedNode.PrimitiveTN(Primitive.Bool, false)
+      case timestamp @ (node, Primitive.Timestamp) =>
+        val format = traits
+          .collectFirst { case t: TimestampFormatTrait =>
+            t.getFormat()
+          }
+          .getOrElse(TimestampFormatTrait.Format.DATE_TIME)
```

Comment:
**issue:** the default should probably be `EPOCH_SECONDS`, it seems to be the case in e.g. AWS's protocol compliance tests.

Reply:
Changed, for some reason I was sure that the `date-time` is default, but that's incorrect, just verified with encoders.

Reply:
it used to be :)